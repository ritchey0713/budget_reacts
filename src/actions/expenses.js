import uuid from "uuid"
import database from "../firebase/firebase"

// add expense 
export const addExpense = (expense) => {
  return {
    type: "ADD_EXPENSE",
    expense  
  }
}

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "", 
      note = "", 
      amount = 0, 
      createdAt = 0
    } = expenseData
    const expense = {description, note, amount, createdAt}
    database.ref("expenses").push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })
  }
}

export const removeExpense = ({ id } = {}) => {
  return {
    type: "REMOVE_EXPENSE",
    id
  }
}

export const editExpense = (id, updates) => {
  return {
    type: "EDIT_EXPENSE",
    id, 
    updates
  }
}


//SET expenses
export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
};

export const startSetExpenses = () => {
 return (dispatch) => {
   return database.ref("expenses").once("value")
    .then((snapShot) => {
      const expenses = []
      snapShot.forEach((childSnapShot) => {
        expenses.push({
          id: childSnapShot.key,
          ...childSnapShot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
 }
};

// export const startSetExpenses = () => {
//   return (dispatch) => {
//     return database.ref("expenses").on("value", (snapShot) => {
//       const expenses = []
//       snapShot.forEach((childSnapShot) => {
//         return expenses.push({
//           id: childSnapShot.key,
//           ...childSnapShot.val()
//         })
//       })
//       dispatch(setExpenses(expenses))
//     })
//   }
// }
