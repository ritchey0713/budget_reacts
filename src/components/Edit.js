import React, { Component } from 'react'
import ExpenseForm from "./ExpenseForm"
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpense extends Component {

  reduxOnSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push("/")
  }
  
  reduxOnDelete = () => {
    this.props.startRemoveExpense(this.props.expense)
    this.props.history.push("/")
  }
  
  render(){
    return (
      <div>
      <div className="page-header">
        <div className="content-container">
        <h1 className="page-header__title">Editing: {this.props.expense.description}</h1>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm 
          expense={this.props.expense}
          reduxOnSubmit={this.reduxOnSubmit}
        />
        <button 
          className="button button--secondary"
          onClick={this.reduxOnDelete}>
          Remove
        </button>
      </div> 
    </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === ownProps.match.params.id
    })
  }
}

const mapDispatchToProps = {  startEditExpense,
  startRemoveExpense}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpense)