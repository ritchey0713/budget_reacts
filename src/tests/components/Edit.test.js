import React from "react"
import { shallow } from "enzyme"
import { EditExpense } from "../../components/Edit.js"
import expenses from "../fixtures/expenses"

let onSubmit, history, wrapper, onDelete;

beforeEach(() => {
  onSubmit = jest.fn()
  onDelete = jest.fn()
  history = {push: jest.fn()}
  wrapper = shallow(<EditExpense expense={expenses[0]} startEditExpense={onSubmit} startRemoveExpense={onDelete} history={history}/>)
})

test("Should render edit page", () => {
  expect(wrapper).toMatchSnapshot()
})

test("should handle startEditExpense", () => {
  wrapper.find("ExpenseForm").prop("reduxOnSubmit")(expenses[0])
  expect(history.push).toHaveBeenLastCalledWith("/")
  expect(onSubmit).toHaveBeenLastCalledWith(expenses[0].id, expenses[0])
})

test("should handle onDelete event", () => {
  wrapper.find("button").simulate("click")
  expect(history.push).toHaveBeenLastCalledWith("/")
  expect(onDelete).toHaveBeenLastCalledWith(expenses[0])
})