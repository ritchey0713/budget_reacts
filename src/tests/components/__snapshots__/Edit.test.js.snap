// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit page 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Editing: 
        coffee
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 1000000,
          "createdAt": 0,
          "description": "coffee",
          "id": 1,
          "note": "",
        }
      }
      reduxOnSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
